### DISCUSSION TITLE
ALL the aws sdk get the device time from this API(clock_gettime), right?

### DISCUSSION THREAD

**Discussion by pppaulpeter** [aws/aws-sdk-cpp#2267] [Q&A] [answered]
### **Unfortunately I have a linux camera which when i call the gettimeofday, it will return local time not UTC time. That is why i need modify the aws sdk code to change the local time to UTC time. otherwise the aws cloud will reset the connection since the time is not sync. So the following code is the only place i need to modify, right?** 

The aws sdk API definition is here:
```cpp
_int aws_sys_clock_get_ticks(uint64_t *timestamp) {
    int ret_val = 0;

    struct timespec ts;
    ret_val = clock_gettime(CLOCK_REALTIME, &ts);
    if (ret_val) {
        return aws_raise_error(AWS_ERROR_CLOCK_FAILURE);
    }

    uint64_t secs = (uint64_t)ts.tv_sec;
    uint64_t n_secs = (uint64_t)ts.tv_nsec;
    *timestamp = (secs * NS_PER_SEC) + n_secs;

    return AWS_OP_SUCCESS;
}_
```

**Comment by jmklix** [comment, marked as answer]
This should be handled automatically for you by the sdk so you don't need to make any changes. This was fixed back in [2016](https://github.com/aws/aws-sdk-cpp/issues/285#issuecomment-251577061):
>I just pushed a fix for this. The way it works is as follows. If we recieve a signing error, we check the date header returned from the server. If it's off, we automatically adjust the clock skew and retry the request. Once it is adjusted the first time, the client will have the correct timestamp. This should happen invisibly to the user without any errors surfacing.

This is enabled with the `enableClockSkewAdjustment` in the client configuration, but this defaults to true. So you don't need to make any changes. Here is the code that makes the changes if you are interested in how it works: [AdjustClockSkew](https://github.com/aws/aws-sdk-cpp/blob/cd22de13474d8f56d75243c613be11957c47164f/aws-cpp-sdk-core/source/client/AWSClient.cpp#L198-L231)

**Reply by pppaulpeter** [reply]
Hi Jmklix,

great! thanks for the quick response!
 i just check the code, so the fix is not try to sync the device time with the cloud server time, right? event the device time is not sync with the server, the authentification will not failed, right? 
how about the aws webrtc library? 
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c
Does it have the same fix? 
this is the log when i run the kvsWebrtcClientMaster sample on the camera device
![image](https://user-images.githubusercontent.com/24501729/210685278-1632f5dd-0b71-4efc-a21d-9150760bfb9f.png)

**Reply by jmklix** [reply]
Correct, this sdk doesn't change the device time. But if the device time is correct then this wouldn't happen. It will still fail once, but then with the retry it will get the server time from the previous error.

As for webrtc, it does not look like it has anything similar. You can ask for more details on that repo

**Comment by pppaulpeter** [comment]
Hi Jmklix,

ok. thanks for your support!
